plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-2-field.tsx

```diff
@@ -0,0 +1,98 @@
+/**
+ * External dependencies
+ */
+import { ValidatedTextInput } from '@woocommerce/blocks-components';
+import { AddressFormValues, ContactFormValues } from '@woocommerce/settings';
+import {
+	useCallback,
+	useEffect,
+	useMemo,
+	useState,
+	Fragment,
+} from '@wordpress/element';
+import { objectHasProp } from '@woocommerce/types';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { AddressFieldProps } from './types';
+
+const Address2Field = < T extends AddressFormValues | ContactFormValues >( {
+	field,
+	fieldProps,
+	fieldsRef,
+	onChange,
+	values,
+}: AddressFieldProps< T > ): JSX.Element => {
+	const hasFieldValue = useMemo( () => {
+		return objectHasProp( values, 'address_2' ) && values.address_2 !== '';
+	}, [ values ] );
+
+	const [ hasFieldBeenModified, setHasFieldBeenModified ] = useState( false );
+
+	const isFieldRequired = field ? field.required : false;
+
+	const [ isFieldVisible, setFieldVisible ] = useState(
+		hasFieldValue || isFieldRequired
+	);
+
+	const toggleFieldVisibility = useCallback(
+		( event: React.MouseEvent< HTMLElement > ) => {
+			event.preventDefault();
+			setFieldVisible( ( prevVisibility: boolean ) => ! prevVisibility );
+		},
+		[]
+	);
+
+	useEffect( () => {
+		/**
+		 * Determine if the address 2 input field should be shown based on the following conditions:
+		 *
+		 * 1. If the address 2 field is required.
+		 * 2. If the address 2 field has a value.
+		 * 3. If the seller has edited the address 2 field.
+		 */
+		const shouldShowField =
+			isFieldRequired || hasFieldValue || hasFieldBeenModified;
+
+		setFieldVisible( shouldShowField );
+	}, [ isFieldRequired, hasFieldValue, hasFieldBeenModified ] );
+
+	return (
+		<Fragment key={ field.key }>
+			{ isFieldVisible ? (
+				<ValidatedTextInput
+					{ ...fieldProps }
+					key={ field.key }
+					ref={ ( el ) => ( fieldsRef.current[ field.key ] = el ) }
+					type={ field.type }
+					label={
+						isFieldRequired ? field.label : field.optionalLabel
+					}
+					className={ `wc-block-components-address-form__${ field.key }` }
+					value={ values[ field.key ] }
+					onChange={ ( newValue: string ) => {
+						onChange( {
+							...values,
+							[ field.key ]: newValue,
+						} );
+						setHasFieldBeenModified( true );
+					} }
+				/>
+			) : (
+				<button
```

Comment:
One tricky note, this now doesn't support browser autofill, which is a tricky issue. To not block this, but I still think it's important to do, is to open another follow up after this.

One way we did autofill before was to render a hidden input (not sure if type hidden or just visually hidden) that would capture the autofill value, if captured, it will expand the actual field and fill it with that value.

Reply:
Thanks for mentioning this, @senadir. I created #47586 for following up on this.